Enable WASM on Windows

iangrunert opened the pull request:
#### 937fd6bd04891f1849a27eb5ff4ac5ee7a3eaeaf
<pre>
Enable WASM on Windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=222315">https://bugs.webkit.org/show_bug.cgi?id=222315</a>

Reviewed by Justin Michaud.

Enable the WASM LLInt on Windows. We use csr2 in place of ws1 as a
scratch register inside WebAssembly.asm on Windows, which needs to be
saved / restored on the stack.

A number of changes were required in offlineasm x86.rb for intel syntax
output.

Integer division / remainder operations had issues, which could be
caught by static asserts <a href="https://bugs.webkit.org/show_bug.cgi?id=203692">https://bugs.webkit.org/show_bug.cgi?id=203692</a>

g_wtfConfigForLLInt is used to reference global wtf config, working
around global loading bug <a href="https://bugs.webkit.org/show_bug.cgi?id=175104">https://bugs.webkit.org/show_bug.cgi?id=175104</a>

* .gitignore:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/jit/GPRInfo.h:
* Source/JavaScriptCore/jit/RegisterSet.cpp:
(JSC::RegisterSetBuilder::wasmPinnedRegisters):
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
(JSC::LLInt::inPlaceInterpreterEntryThunk):
(JSC::LLInt::inPlaceInterpreterEntryThunkSIMD):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/llint/WebAssembly64.asm:
* Source/JavaScriptCore/offlineasm/asm.rb:
* Source/JavaScriptCore/offlineasm/x86.rb:
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/JavaScriptCore/runtime/JSCConfig.cpp:
* Source/JavaScriptCore/runtime/JSCConfig.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::LLIntCallee::calleeSaveRegistersImpl):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmStreamingParser.cpp:
(JSC::Wasm::StreamingParser::addBytes):
* Source/cmake/OptionsJSCOnly.cmake:
* Source/cmake/OptionsWin.cmake:
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: <a href="https://commits.webkit.org/270429@main">https://commits.webkit.org/270429@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/830864930839072fc62a3693a87c8c9bc5ac2330

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25494 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27597 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23364 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25773 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5823 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1521 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23523 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25739 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3018 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21979 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28178 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2669 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22928 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29032 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22176 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23247 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23291 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26858 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/24720 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2665 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/919 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/32156 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4046 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7005 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6099 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3119 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3004 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

iangrunert commented on Source/JavaScriptCore/runtime/JSCConfig.cpp:
```diff
@@ -25,13 +25,18 @@
 #endif
 
+#if OS(WINDOWS) && ENABLE(WEBASSEMBLY)
+WTF::Config g_wtfConfigForLLInt = g_wtfConfig;
```
This fails check-webkit-style for using an underscore in the identifier name, but I'm following the convention of `g_jscConfig`.

justinmichaud replied:
I am not so sure this is right. Why is this needed?

iangrunert replied:
This is due to a limitation in offlineasm, which can't resolve the g_wtfConfig label from the DLL for WTF. By resolving the import on the C++ side, the g_wtfConfigForLLInt label is in the JSC DLL and can be resolved.

https://bugs.webkit.org/show_bug.cgi?id=175104

justinmichaud replied:
I am still not sure this is correct? Shouldn't it be a reference then? Here, it seems like we copy the config. Anyway, since it is Windows-only, I won't block landing the patch on this.

justinmichaud commented on Source/JavaScriptCore/jit/GPRInfo.h:
```diff
@@ -483,8 +483,13 @@ class GPRInfo {
     static constexpr GPRReg wasmScratchGPR1 = X86Registers::r10;
 #endif
     static constexpr GPRReg wasmContextInstancePointer = regCS0;
+#if !OS(WINDOWS)
```
Can we swap these cases?

iangrunert replied:
It's following the same convention as the rest of the Windows preprocessor directives in this file, should I swap them all?

justinmichaud commented on Source/JavaScriptCore/jit/RegisterSet.cpp:
```diff
@@ -448,6 +448,9 @@ RegisterSet RegisterSetBuilder::wasmPinnedRegisters()
     if constexpr (GPRInfo::wasmBoundsCheckingSizeRegister != InvalidGPRReg)
         result.add(GPRInfo::wasmBoundsCheckingSizeRegister, IgnoreVectors);
+#if OS(WINDOWS)
+    result.add(GPRInfo::regCS2, IgnoreVectors);
```
Can we give this a symbolic name?

justinmichaud commented on Source/JavaScriptCore/llint/LLIntThunks.cpp:
```diff
@@ -227,7 +227,11 @@ MacroAssemblerCodeRef<JITThunkPtrTag> inPlaceInterpreterEntryThunk()
     std::call_once(onceKey, [&] {
         JSInterfaceJIT jit;
         void* ptr = reinterpret_cast<void*>(ipint_entry);
+#if COMPILER(MSVC)
```
can we wrap this workaround in a function?

rkirsling commented on Source/JavaScriptCore/offlineasm/x86.rb:
```diff
@@ -70,13 +70,13 @@
-# r13 =>             csr4 (callee-save)
-# r14 =>             csr5 (callee-save, numberTag)
-# r15 =>             csr6 (callee-save, notCellMask)
+# rdi =>    ws1, t6, csr2 (callee-save, wasmScratch)
```
nit: 
```suggestion
# rdi => t6,         csr2 (callee-save, wasmScratch)
```

rkirsling reviewed (commented):
I feel like I shouldn't give the final approval here but it's looking good from my perspective.

justinmichaud commented on Source/JavaScriptCore/llint/WebAssembly64.asm:
```diff
@@ -386,7 +465,12 @@ wasmOp(i64_rem_u, WasmI64RemU, macro (ctx)
-    returnq(ctx, t2)
+
+    if X86_64_WIN
+        returnq(ctx, t1)
```
can we just give t1/t2 a meaningful name? And maybe above, we should be using the hardware names so that this won't break if we change the names in the future? If this isn't possible that's fine

justinmichaud replied:
Hmm, I don't know exactly how, but I feel like this can be better. A bunch of these cases seem like they would be redundant with a good local register name plus some real hardware register names

iangrunert replied:
Yeah I can give that a try, should be able to pull the register definition to a const and then re-use the same implementation for these division / remainder ops.

iangrunert replied:
Gave this a shot, pushed in a second commit. It does reduce some duplication, though I'm not 100% sure whether to extend the variables into the ARM / RISCV branches.

In LowLevelInterpreter we `emit` direct assembly, but to do that you have to emit in Intel syntax for Windows and AT&T for everything else. It's a bit of a shame - that Intel vs. AT&T syntax also introduces a bunch of duplication in offlineasm, which could get eliminated if we used Intel syntax on the non-Windows platform linker (or figure out how to switch from MASM to clang and enable AT&T syntax on Windows).

justinmichaud reviewed (commented):
Overall this looks good to me. I would love one more quick pass to see if we can reduce some duplication, and then it should be ready to land

justinmichaud reviewed (approved):
r=me

iangrunert committed:
Enable WASM on Windows
https://bugs.webkit.org/show_bug.cgi?id=222315

Reviewed by Justin Michaud.

Enable the WASM LLInt on Windows. We use csr2 in place of ws1 as a
scratch register inside WebAssembly.asm on Windows, which needs to be
saved / restored on the stack.

A number of changes were required in offlineasm x86.rb for intel syntax
output.

Integer division / remainder operations had issues, which could be
caught by static asserts https://bugs.webkit.org/show_bug.cgi?id=203692

g_wtfConfigForLLInt is used to reference global wtf config, working
around global loading bug https://bugs.webkit.org/show_bug.cgi?id=175104

* .gitignore:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/jit/GPRInfo.h:
* Source/JavaScriptCore/jit/RegisterSet.cpp:
(JSC::RegisterSetBuilder::wasmPinnedRegisters):
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
(JSC::LLInt::inPlaceInterpreterEntryThunk):
(JSC::LLInt::inPlaceInterpreterEntryThunkSIMD):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/llint/WebAssembly64.asm:
* Source/JavaScriptCore/offlineasm/asm.rb:
* Source/JavaScriptCore/offlineasm/x86.rb:
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/JavaScriptCore/runtime/JSCConfig.cpp:
* Source/JavaScriptCore/runtime/JSCConfig.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::LLIntCallee::calleeSaveRegistersImpl):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmStreamingParser.cpp:
(JSC::Wasm::StreamingParser::addBytes):
* Source/cmake/OptionsJSCOnly.cmake:
* Source/cmake/OptionsWin.cmake:
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: https://commits.webkit.org/270429@main
